go/storage/mkvs/db/badger/migrate.go

```diff
@@ -0,0 +1,409 @@
+			}
+
+		case v3RootUpdatedNodesKeyFmt.Decode(key, &version, &h1):
+			var oldUpdatedNodes []v3UpdatedNode
+			err := it.Item().Value(func(data []byte) error {
+				return cbor.UnmarshalTrusted(data, &oldUpdatedNodes)
+			})
+			if err != nil {
+				return 0, fmt.Errorf("mkvs/badger/migrate: error decoding updated nodes list for root %v:%v: %w", h1, version, err)
+			}
+
+			root, err := v4.helper.GetRootForHash(h1, version)
+			if err != nil {
+				return 0, fmt.Errorf("mkvs/badger/migrate: error getting root %v:%v for updated nodes list: %w", h1, version, err)
+			}
+			v4.roots[typedHashFromRoot(root)] = struct{}{}
+
+			newUpdatedNodes := make([]v4UpdatedNode, 0, len(oldUpdatedNodes))
+			for _, up := range oldUpdatedNodes {
+				newUpdatedNodes = append(newUpdatedNodes, v4UpdatedNode{
+					Removed: up.Removed,
+					Hash:    typedHashFromParts(root.Type, up.Hash),
+				})
+			}
+
+			entry := badger.NewEntry(
+				v4RootUpdatedNodesKeyFmt.Encode(version, typedHashFromRoot(root)),
+				cbor.Marshal(newUpdatedNodes),
+			)
+			err = v4.changeBatch.SetEntryAt(entry, it.Item().Version())
+			if err != nil {
+				return 0, fmt.Errorf("mkvs/badger/migrate: error storing updated nodes list for root %v: %w", root, err)
+			}
+			err = v4.changeBatch.DeleteAt(key, it.Item().Version())
+			if err != nil {
+				return 0, fmt.Errorf("mkvs/badger/migrate: error deleting old nodes nodes list for root %v: %w", root, err)
+			}
+
+		case v3MetadataKeyFmt.Decode(key):
+			// Nothing to do here for v3->v4 migration, no hashes stored in the metadata.
+
+		case v3MultipartRestoreNodeLogKeyFmt.Decode(key, &h1):
+			// Tree nodes will be updated once we have all roots.
+		}
+
+		currentCount++
+		v4.helper.ReportProgress("updated keys", currentCount, metaCount)
+	}
+
+	v4.helper.ReportStatus("migrating tree nodes")
+	v4.currentTreeCount = 0
+	for root := range v4.roots {
+		if err := v4.walkTree(root.Type(), root.Hash()); err != nil {
+			return 0, fmt.Errorf("mkvs/badger/migrate: error walking node tree under %v: %w", root, err)
+		}
+	}
+
+	err := v4.changeBatch.Flush()
+	if err != nil {
+		return 0, fmt.Errorf("mkvs/badger/migrate: error writing changes to the database: %w", err)
+	}
+	return 4, nil
+}
+
+func Migrate(cfg *api.Config, helper MigrationHelper) (uint64, error) {
+	db := &badgerNodeDB{
+		logger:           logging.GetLogger("mkvs/db/badger/migrate"),
+		namespace:        cfg.Namespace,
+		discardWriteLogs: cfg.DiscardWriteLogs,
+	}
+
+	opts := badger.DefaultOptions(cfg.DB)
+	opts = opts.WithLogger(cmnBadger.NewLogAdapter(db.logger))
+	opts = opts.WithSyncWrites(!cfg.NoFsync)
+	// Allow value log truncation if required (this is needed to recover the
+	// value log file which can get corrupted in crashes).
+	opts = opts.WithTruncate(true)
+	opts = opts.WithCompression(options.Snappy)
+	opts = opts.WithBlockCacheSize(cfg.MaxCacheSize)
+	opts = opts.WithDetectConflicts(false)
```

Comment:
Should this part be moved to some common function? E.g. `configToBadgerOptions(cfg)` or similar?